contracts/RewardsStandaloneSingleToken.sol

```diff
@@ -0,0 +1,214 @@
+        uint256 balanceNow = rewardToken.balanceOf(address(this));
+
+        if (balanceNow == 0 || balanceNow <= balanceBefore) {
+            balanceBefore = balanceNow;
+            return;
+        }
+
+        // if there's no bond staked, it doesn't make sense to ackFunds because there's nobody to distribute them to
+        // and the calculation would fail anyways due to division by 0
+        if (poolSize == 0) {
+            return;
+        }
+
+        uint256 diff = balanceNow.sub(balanceBefore);
+        uint256 multiplier = currentMultiplier.add(diff.mul(decimals).div(poolSize));
+
+        balanceBefore = balanceNow;
+        currentMultiplier = multiplier;
+    }
+
+    // setupPullToken is used to setup the rewards system; only callable by contract owner
+    // set source to address(0) to disable the functionality
+    function setupPullToken(address source, uint256 startTs, uint256 endTs, uint256 amount) public {
+        require(msg.sender == owner(), "!owner");
+        require(!disabled, "contract is disabled");
+
+        if (pullFeature.source != address(0)) {
+            require(source == address(0), "contract is already set up, source must be 0x0");
+            disabled = true;
+        } else {
+            require(source != address(0), "contract is not setup, source must be != 0x0");
+        }
+
+        if (source == address(0)) {
+            require(startTs == 0, "disable contract: startTs must be 0");
+            require(endTs == 0, "disable contract: endTs must be 0");
+            require(amount == 0, "disable contract: amount must be 0");
+        } else {
+            require(endTs > startTs, "setup contract: endTs must be greater than startTs");
+            require(amount > 0, "setup contract: amount must be greater than 0");
+        }
+
+        pullFeature.source = source;
+        pullFeature.startTs = startTs;
+        pullFeature.endTs = endTs;
+        pullFeature.totalDuration = endTs.sub(startTs);
+        pullFeature.totalAmount = amount;
+
+        if (lastPullTs < startTs) {
+            lastPullTs = startTs;
+        }
+    }
+
+    // _pullToken calculates the amount based on the time passed since the last pull relative
+    // to the total amount of time that the pull functionality is active and executes a transferFrom from the
+    // address supplied as `pullTokenFrom`, if enabled
+    function _pullToken() internal {
+        if (
+            pullFeature.source == address(0) ||
+            block.timestamp < pullFeature.startTs
+        ) {
+            return;
+        }
+
+        uint256 timestampCap = pullFeature.endTs;
+        if (block.timestamp < pullFeature.endTs) {
+            timestampCap = block.timestamp;
+        }
+
+        if (lastPullTs >= timestampCap) {
+            return;
+        }
+
+        uint256 timeSinceLastPull = timestampCap.sub(lastPullTs);
+        uint256 shareToPull = timeSinceLastPull.mul(decimals).div(pullFeature.totalDuration);
```

Comment:
I think we should change the calculation here to include a `rewardPerSecond`. This way we can increase or decrease `rewardPerSecond` any time we want and there is no need for `totalAmount` anymore.